### PR TITLE
packages: allow-/block-list package repos [backend]

### PR DESCRIPTION
**Pull request by Strum355**
Implements allow-/block-listing of package repos and versions. Components include:
1. the service+store layers
2. a background job to calculate new blocked statuses for everything when any filters change
3. graphql layer for the :sparkles: UI :sparkles: 

## Test plan

Added unit tests for the background jobs, ran graphql queries and mutations locally numerous times, cross-checked with running sql queries manually, and ran full-flow locally with uploads, adding/changing/deleting filters etc

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.01% (+0.40 kb) | 0.01% (+1.70 kb) | 0.01% (+1.30 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/nsc/allowblocklist-packages/compare-report.html;tab=live_object) for a full comparison between the commits 8b0f899 and 6ffd017 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: enterprise/cmd/worker/shared/shared.go [hunk 1]
```diff
@@ -59,6 +59,7 @@ var additionalJobs = map[string]job.Job{
 	"codeintel-uploadstore-expirer":               codeintel.NewPreciseCodeIntelUploadExpirer(),
 	"codeintel-crates-syncer":                     codeintel.NewCratesSyncerJob(),
 	"codeintel-sentinel-cve-scanner":              codeintel.NewSentinelCVEScannerJob(),
+	"codeintel-package-filter-applicator":         codeintel.NewPackagesFilterApplicatorJob(),
```
Comment (efritz) [inline enterprise/cmd/worker/shared/shared.go:62]: Can you also put a note in workers.md?

File: internal/codeintel/dependencies/internal/background/job_packages_filter.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		latestUpdated = filters[0].DeletedAt.Time
+	}
+
+	if needsFiltering, err := j.store.ExistsPackageRepoRefLastCheckedBefore(ctx, latestUpdated); !needsFiltering || err != nil {
+		// returns nil if err is nil, so its fine
+		return errors.Wrap(err, "failed to check whether package repo filters need applying to anything")
+	}
+
+	pkgsAffected, versionsAffected, err = j.store.ApplyPackageFilters(ctx)
+	fmt.Println("DOING THE THING", pkgsAffected, versionsAffected, err)
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/background/job_packages_filter.go:68]: :)

File: internal/codeintel/dependencies/shared/types.go [hunk 3]
```diff
@@ -28,3 +36,30 @@ type MinimialVersionedPackageRepo struct {
...
+	Behaviour     string
+	PackageScheme string
+	NameFilter    *struct {
+		PackageGlob string
+	}
+	VersionFilter *struct {
+		PackageName string
+		VersionGlob string
+	}
+	DeletedAt pq.NullTime
```
Comment (efritz) [inline internal/codeintel/dependencies/shared/types.go:63]: Why is sql stuff leaking from the db layer?

File: internal/codeintel/dependencies/internal/background/job_packages_filter.go (hunk 2 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/background/job_packages_filter.go:68]: need to put some MATRICS here

File: internal/codeintel/dependencies/service.go [hunk 4]
```diff
@@ -87,3 +90,88 @@ func (s *Service) DeletePackageRepoRefVersionsByID(ctx context.Context, ids ...i
...
+
+func deref(s *string) string {
+	if s == nil {
+		return ""
+	}
+	return *s
+}
+
+func (s *Service) ListPackageRepoRefFilters(ctx context.Context, opts ListPackageRepoRefFiltersOpts) (_ []shared.PackageFilter, err error) {
+	ctx, _, endObservation := s.operations.listPackageRepoFilters.With(ctx, &err, observation.Args{LogFields: []log.Field{
```
Comment (efritz) [inline internal/codeintel/dependencies/service.go:109]: wrong operation

File: internal/codeintel/dependencies/shared/types.go (hunk 3 above)
Comment (Strum355) [reply internal/codeintel/dependencies/shared/types.go:63]: coz it made one line very convenient given zero value of time.Time lol

File: internal/codeintel/dependencies/service.go (hunk 4 above)
Comment (Strum355) [reply internal/codeintel/dependencies/service.go:109]: Correct one, just marginally different names. WIll consolidate

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 5]
```diff
@@ -3654,6 +3707,50 @@ type PackageRepoReferenceVersion {
...
+}
+
+type PackageVersionFilter {
+    packageName: String!
+
+    versionGlob: String!
+}
+
+"""
+"""
```
Comment (efritz) [inline cmd/frontend/graphqlbackend/schema.graphql:3738]: I'd fill it out or remove it, but just a reminder there's a lot of blank docs (a lot of fields seem explanatory though).

Comment (Strum355) [reply cmd/frontend/graphqlbackend/schema.graphql:3738]: Theyll all be filled out yea

File: internal/codeintel/dependencies/internal/background/job_packages_filter.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+			return a.DeletedAt.Time.After(b.DeletedAt.Time)
+		}
+
+		return a.UpdatedAt.After(b.UpdatedAt)
+	})
+
+	latestUpdated := filters[0].UpdatedAt
+	if filters[0].DeletedAt.Valid {
+		latestUpdated = filters[0].DeletedAt.Time
+	}
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/background/job_packages_filter.go:60]: Looks like this entire block only calculates `latestUpdated`, and I think you could push this down to the SQL layer.

Comment (efritz) [reply internal/codeintel/dependencies/internal/background/job_packages_filter.go:60]: You could possibly even encompass what `ExistsPackageRepoRefLastCheckedBefore` does.

File: migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        ) AS m
+    );
+
+    RETURN allowed;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Transformed from Python stdlib fnmatch.translate function
+-- as licensed by https://sourcegraph.com/github.com/python/cpython@54dfa14c5a94b893b67a4d9e9e403ff538ce9023/-/blob/LICENSE
+CREATE OR REPLACE FUNCTION glob_to_regex(pat text) RETURNS text AS $$
```
Comment (efritz) [inline migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql:245]: God help us.

File: internal/codeintel/dependencies/internal/store/store.go [hunk 8]
```diff
@@ -402,3 +430,400 @@ const deleteDependencyRepoVersionsByID = `
...
+	ctx, _, endObservation := s.operations.listPackageRepoFilters.With(ctx, &err, observation.Args{LogFields: []log.Field{
+		log.Int("numPackageRepoFilterIDs", len(opts.IDs)),
+		log.String("packageScheme", opts.PackageScheme),
+		log.Int("after", opts.After),
+		log.Int("limit", opts.Limit),
+	}})
+	defer endObservation(1, observation.Args{})
+
+	conds := make([]*sqlf.Query, 0, 4)
+	// we never want to surface filters, and the table should be small enough that a
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/store/store.go:451]: we never want to surface _deleted_ filters?

File: internal/codeintel/dependencies/internal/store/store.go [hunk 9]
```diff
@@ -402,3 +430,400 @@ const deleteDependencyRepoVersionsByID = `
...
+		conds = append(conds, sqlf.Sprintf("id = ANY(%s)", pq.Array(opts.IDs)))
+	}
+
+	if opts.PackageScheme != "" {
+		conds = append(conds, sqlf.Sprintf("scheme = %s", opts.PackageScheme))
+	}
+
+	if opts.After != 0 {
+		conds = append(conds, sqlf.Sprintf("id > %s", opts.After))
+	}
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/store/store.go:465]: But you're not ordering :clueless:

File: internal/codeintel/dependencies/internal/store/store.go [hunk 10]
```diff
@@ -402,3 +430,400 @@ const deleteDependencyRepoVersionsByID = `
...
+	}})
+	defer endObservation(1, observation.Args{})
+
+	_, exists, err = basestore.ScanFirstInt(s.db.Query(ctx, sqlf.Sprintf(existsPackageRepoRefLastCheckedBeforeQuery, before, before)))
+	return
+}
+
+const existsPackageRepoRefLastCheckedBeforeQuery = `
+SELECT 1
+WHERE EXISTS (
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/store/store.go:609]:
```suggestion
SELECT EXISTS (
```

File: internal/codeintel/dependencies/internal/store/store.go (hunk 8 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/store/store.go:451]: Oops youre right, there is a point where we do want to. Bg job needs to check for this

File: internal/codeintel/dependencies/internal/store/store.go (hunk 9 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/store/store.go:465]: Probably lost to the million changes this underwent

File: internal/codeintel/dependencies/internal/store/store.go (hunk 10 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/store/store.go:609]: Returns 1 or TRUE?

File: migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql (hunk 7 above)
Comment (Strum355) [reply migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql:245]: By the blessings of Cody

File: internal/codeintel/dependencies/internal/background/job_packages_filter.go (hunk 6 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/background/job_packages_filter.go:60]: Yea was considering that second one. Good point on the first too

**Review by efritz — COMMENTED** [review]
I freakin love Postgres (and what you've done with it here).

LGTM from a first glance. Would like to see unit tests for each of the store methods at a minimum. Left comments on things that felt shaky. Can do another cursory in-depth pass post-draft, pre-merge (just ping me).

File: internal/codeintel/dependencies/internal/store/store.go (hunk 8 above)
Comment (efritz) [reply internal/codeintel/dependencies/internal/store/store.go:451]: I just assumed that's what the comment was specifying?

File: migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql (hunk 7 above)
Comment (efritz) [reply migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql:245]: Under His eye.

File: internal/codeintel/dependencies/internal/store/store.go (hunk 8 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/store/store.go]: the comment was wrong, we do want to surface them optionally. Fixed

File: internal/codeintel/dependencies/internal/background/job_packages_filter.go (hunk 6 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/background/job_packages_filter.go]: Done, pls check for correctness

File: migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql (hunk 7 above)
Comment (Strum355) [reply migrations/frontend/1677166643_package_repos_allowblock_lists/up.sql]: RIP Cody's work

**Review by efritz — APPROVED** [review]

File: internal/codeintel/dependencies/internal/background/observability.go [hunk 11]
```diff
@@ -2,22 +2,56 @@ package background
...
+		counter := prometheus.NewCounter(prometheus.CounterOpts{
+			Name: name,
+			Help: help,
+		})
+		observationCtx.Registerer.MustRegister(counter)
+
+		metricsMap[name] = counter
+
+		return counter
+	}
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/background/observability.go:54]: Is this worth putting into a shared package (in a follow-up PR)? Seems like we've done this a few times now.

Comment (Strum355) [reply internal/codeintel/dependencies/internal/background/observability.go:54]: probably, just need to design a good enough abstraction

**Review by efritz — APPROVED** [review]

**Pull request merged**